ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java

```diff
@@ -0,0 +1,215 @@
+    this.blockBroadcaster = blockBroadcaster;
+    this.protocolContext = protocolContext;
+    this.ethPeers = ethPeers;
+    this.syncState = syncState;
+    this.miningEnabled = miningEnabled;
+    this.transactionPool = transactionPool;
+  }
+
+  BlockStats convertToStats(final Block block, final Difficulty difficulty) {
+    return convertToStats(
+        block.getHeader(),
+        block.getBody().getTransactions(),
+        block.getBody().getOmmers(),
+        difficulty);
+  }
+
+  BlockStats convertToStats(
+      final BlockHeader header,
+      final List<Transaction> txs,
+      final List<BlockHeader> ommers,
+      final Difficulty difficulty) {
+    return new BlockStats(
+        UInt256.valueOf(header.getNumber()),
+        Hash.fromBytes(header.getHash()),
+        Hash.fromBytes(header.getParentHash()),
+        header.getTimestamp(),
+        Address.fromBytes(header.getCoinbase()),
+        header.getGasUsed(),
+        header.getGasLimit(),
+        header.getDifficulty().toUInt256(),
+        header.getDifficulty().add(difficulty).toUInt256(),
+        txs.stream().map(this::convertTxToStats).collect(Collectors.toList()),
+        Hash.fromBytes(header.getTransactionsRoot()),
+        Hash.fromBytes(header.getStateRoot()),
+        ommers.stream().map(this::convertToBlockHeader).collect(Collectors.toList()));
+  }
+
+  private org.apache.tuweni.eth.BlockHeader convertToBlockHeader(final BlockHeader header) {
+    return new org.apache.tuweni.eth.BlockHeader(
+        Hash.fromBytes(header.getParentHash()),
+        Hash.fromBytes(header.getOmmersHash()),
+        Address.fromBytes(header.getCoinbase()),
+        Hash.fromBytes(header.getStateRoot()),
+        Hash.fromBytes(header.getTransactionsRoot()),
+        Hash.fromBytes(header.getReceiptsRoot()),
+        header.getLogsBloom(),
+        header.getDifficulty().toUInt256(),
+        UInt256.valueOf(header.getNumber()),
+        Gas.valueOf(header.getGasLimit()),
+        Gas.valueOf(header.getGasUsed()),
+        Instant.ofEpochMilli(header.getTimestamp()),
+        header.getExtraData(),
+        Hash.fromBytes(header.getMixHash()),
+        Bytes.ofUnsignedLong(header.getNonce()));
+  }
+
+  private TxStats convertTxToStats(final Transaction transaction) {
+    return new TxStats(Hash.fromBytes(transaction.getHash()));
+  }
+
+  private void onNewBlock(final Block block, final Difficulty difficulty) {
+    reporter.sendNewHead(
+        convertToStats(
+            block.getHeader(),
+            block.getBody().getTransactions(),
+            block.getBody().getOmmers(),
+            difficulty));
+  }
+
+  public void start(final Vertx vertx) {
+    Log4j2LoggerProvider provider = new Log4j2LoggerProvider();
+    URI ethStatsURI =
+        URI.create(
+            "ws://" + ethStatsParameters.getHost() + ":" + ethStatsParameters.getPort() + "/api");
+    // TODO provide client identification string
+    this.reporter =
+        new EthStatsReporter(
```

Comment:
I'm not terribly excited about using an external library for what should be core functionality.  Would this make more sense as a third party plugin via the 1.4 plugins APIs?  Is there any stats we need to expose that are not already available via the plugin?

Reply:
No idea, but I'm pretty sure there are quite a few things that you'd need to report, such as the mining hashrate, which are not currently exposed.

Maybe this is a good stopping point: either you want this in or as a plugin. What should it be?

Reply:
I have moved this discussion to rocket.chat to a wider audience, we can discuss more there.